### ISSUE TITLE
Use TensorFlow public APIs in seq2seq module

### ISSUE THREAD

**Issue by guillaumekln** [tensorflow/addons#134]
As mentioned in #72, the seq2seq module should use TensorFlow public APIs. This issue is to track progress on this task.

* [ ] attention_wrapper.py
* [ ] attention_wrapper_test.py
* [ ] basic_decoder.py
* [ ] basic_decoder_test.py
* [ ] beam_search_decoder.py
* [ ] beam_search_decoder_test.py
* [ ] beam_search_ops_test.py
* [ ] decoder.py
* [ ] decoder_test.py
* [x] loss.py
* [x] loss_test.py
* [x] sampler.py

This should not be difficult so people could contribute.

**Comment by armando-fandango** [comment]
@seanpmorgan started working on yesterday since it seems little more complicated to be good first issue.

**Comment by seanpmorgan** [comment]
Thanks @armando-fandango see this PR as an example:
https://github.com/tensorflow/addons/pull/55

**Comment by armando-fandango** [comment]
@guillaumekln @qlzh727 In https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/decoder_test.py on line 65 and 142, my_decoder object is being called, but it doesn't look like the base class hierarchy is made callable, though there is a call() function with almost same signature as in line 65/142.. so was that function supposed to be __call__ ? Thought of asking before I dig deeper to fix it.

**Comment by armando-fandango** [comment]
I remember there was a discussion on this before... did we finally agree to use distributions from tfp ?

**Comment by seanpmorgan** [comment]
> I remember there was a discussion on this before... did we finally agree to use distributions from tfp ?
No we do not plan on depending on tfp. Which imports are you referring to?

**Comment by guillaumekln** [comment]
> @guillaumekln @qlzh727 In https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/decoder_test.py on line 65 and 142, `my_decoder` object is being called, but it doesn't look like the base class hierarchy is made callable, though there is a `call()` function with almost same signature as in line 65/142.. so was that function supposed to be `__call__` ? Thought of asking before I dig deeper to fix it.

It is callable: `BasicDecoder` extends `BaseDecoder` which extends `keras.layers.Layer` which itself implements `__call__`.

See also the recommended way to subclass `Layer`:

https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/layers/Layer

**Comment by armando-fandango** [comment]
> > @guillaumekln @qlzh727 In  
> It is callable: `BasicDecoder` extends `BaseDecoder` which extends `keras.layers.Layer` which itself implements `__call__`.
> 
> See also the recommended way to subclass `Layer`:
> 
> https://www.tensorflow.org/versions/r2.0/api_docs/python/tf/keras/layers/Layer
pylint keeps throwing these errors.

`tensorflow_addons/seq2seq/decoder_test.py:63: [E1102(not-callable), DecodeRNNTest._testDecodeRNN] my_decoder is not callable

tensorflow_addons/seq2seq/decoder_test.py:140: [E1102(not-callable), DecodeRNNTest._testDynamicDecodeRNNWithTrainingHelperMatchesDynamicRNN] my_decoder is not callable`

submitting pull request.

**Comment by armando-fandango** [comment]
> > I remember there was a discussion on this before... did we finally agree to use distributions from tfp ?
> 
> No we do not plan on depending on tfp. Which imports are you referring to?

Sorry my bad, I was looking at some other code in parallel that was using distributions.
